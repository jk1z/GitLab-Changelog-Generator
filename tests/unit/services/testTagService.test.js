const TagService = require("../../../app/services/tag");

const GitlabRepository = require("../../../app/repositories/gitlab");
const LoggerService = require("../../../app/services/logger");
jest.mock("../../../app/repositories/gitlab");
jest.mock("../../../app/services/logger");

const TagFixtures = require("../../fixtures/tag");
const ProjectFixtures = require("../../fixtures/project");

describe("Tag service", () => {
    describe("#getLatestAndSecondLatestTag", () => {
        let tagService;
        let gitlabRepository;
        let loggerService;
        let res;
        describe("Sunny scenario", () => {
            beforeAll(async () => {
                jest.resetAllMocks();
                loggerService = new LoggerService({});
                gitlabRepository = new GitlabRepository({});
                gitlabRepository.findTagsByProjectId.mockResolvedValue({
                    tags: [TagFixtures.tagDefault(), TagFixtures.tagDefaultOther(), TagFixtures.tagDefaultOther2()],
                    _link: jest.fn()
                });
                gitlabRepository.findRepoByProjectId.mockResolvedValue(ProjectFixtures.projectDefault());
                gitlabRepository.findBranchRefsByProjectIdAndSha.mockResolvedValue([{ name: "master" }]);
                tagService = new TagService({
                    gitlabRepository,
                    loggerService,
                    config: {
                        TARGET_BRANCH: "master",
                        TARGET_TAG_REGEX: new RegExp(/^v[0-9]+.[0-9]+.[0-9]+$/),
                        GITLAB_PROJECT_ID: "123456"
                    }
                });
                res = await tagService.getLatestAndSecondLatestTag();
            });
            test("should log info", () => {
                expect(loggerService.info).toBeCalledTimes(2);
                expect(loggerService.info).toMatchSnapshot();
            });
            test("should return latest and second latest tag", () => {
                expect(res).toHaveLength(2);
                expect(res).toMatchSnapshot();
            });
            test("should call findTagsByProjectId", () => {
                expect(gitlabRepository.findTagsByProjectId).toBeCalledTimes(1);
                expect(gitlabRepository.findTagsByProjectId.mock.calls[0]).toMatchSnapshot();
            });
            test("should call findBranchRefsByProjectIdAndSha", () => {
                expect(gitlabRepository.findBranchRefsByProjectIdAndSha).toBeCalledTimes(2);
                expect(gitlabRepository.findBranchRefsByProjectIdAndSha.mock.calls).toMatchSnapshot();
            });
        });
        describe("Optional scenario", () => {
            describe("Only one tag", () => {
                beforeAll(async () => {
                    jest.resetAllMocks();
                    loggerService = new LoggerService({});
                    gitlabRepository = new GitlabRepository({});
                    gitlabRepository.findTagsByProjectId.mockResolvedValue({
                        tags: [TagFixtures.tagDefaultOther2()],
                        _link: jest.fn()
                    });
                    gitlabRepository.findRepoByProjectId.mockResolvedValue(ProjectFixtures.projectDefault());
                    gitlabRepository.findBranchRefsByProjectIdAndSha.mockResolvedValue([{ name: "master" }]);
                    tagService = new TagService({
                        gitlabRepository,
                        loggerService,
                        config: {
                            TARGET_BRANCH: "master",
                            TARGET_TAG_REGEX: new RegExp(/^v[0-9]+.[0-9]+.[0-9]+$/),
                            GITLAB_PROJECT_ID: "123456"
                        }
                    });
                    res = await tagService.getLatestAndSecondLatestTag();
                });
                test("should log info", () => {
                    expect(loggerService.info).toBeCalledTimes(2);
                    expect(loggerService.info).toMatchSnapshot();
                });
                test("should return latest and second latest tag", () => {
                    expect(res).toHaveLength(2);
                    expect(res).toMatchSnapshot();
                });
                test("should call findTagsByProjectId", () => {
                    expect(gitlabRepository.findTagsByProjectId).toBeCalledTimes(1);
                    expect(gitlabRepository.findTagsByProjectId.mock.calls[0]).toMatchSnapshot();
                });
                test("should call findRepoByProjectId", () => {
                    expect(gitlabRepository.findRepoByProjectId).toBeCalledTimes(1);
                    expect(gitlabRepository.findRepoByProjectId.mock.calls[0]).toMatchSnapshot();
                });
                test("should call findBranchRefsByProjectIdAndSha", () => {
                    expect(gitlabRepository.findBranchRefsByProjectIdAndSha).toBeCalledTimes(1);
                    expect(gitlabRepository.findBranchRefsByProjectIdAndSha.mock.calls).toMatchSnapshot();
                });
            });
        });
        describe("Boom scenario", () => {
            let err;
            describe("Latest tag is not matching the regex", () => {
                beforeAll(async () => {
                    jest.resetAllMocks();
                    loggerService = new LoggerService({});
                    gitlabRepository = new GitlabRepository({});
                    gitlabRepository.findTagsByProjectId.mockResolvedValue({
                        tags: [TagFixtures.tagDefault(), TagFixtures.tagDefaultOther(), TagFixtures.tagDefaultOther2()],
                        _link: jest.fn()
                    });
                    gitlabRepository.findBranchRefsByProjectIdAndSha.mockResolvedValue([{ name: "develop" }]);
                    tagService = new TagService({
                        gitlabRepository,
                        loggerService,
                        config: {
                            TARGET_TAG_REGEX: new RegExp(/^[0-9]+.[0-9]+.[0-9]+$/),
                            GITLAB_PROJECT_ID: "123456"
                        }
                    });
                    try {
                        res = await tagService.getLatestAndSecondLatestTag();
                    } catch (e) {
                        err = e;
                    }
                });
                it("should throw err", () => {
                    expect(err.message).toMatchSnapshot();
                });
            });
            describe("Latest tag is not in the targeting branch", () => {
                beforeAll(async () => {
                    jest.resetAllMocks();
                    loggerService = new LoggerService({});
                    gitlabRepository = new GitlabRepository({});
                    gitlabRepository.findTagsByProjectId.mockResolvedValue({
                        tags: [TagFixtures.tagDefault(), TagFixtures.tagDefaultOther(), TagFixtures.tagDefaultOther2()],
                        _link: jest.fn()
                    });
                    gitlabRepository.findBranchRefsByProjectIdAndSha.mockResolvedValue([{ name: "master" }]);
                    tagService = new TagService({
                        gitlabRepository,
                        loggerService,
                        config: {
                            TARGET_BRANCH: "develop",
                            GITLAB_PROJECT_ID: "123456"
                        }
                    });
                    try {
                        res = await tagService.getLatestAndSecondLatestTag();
                    } catch (e) {
                        err = e;
                    }
                });
                it("should throw err", () => {
                    expect(err.message).toMatchSnapshot();
                });
            });
        });
    });
    describe("#isTagsMatchTargetTagRegex", () => {
        let tagService;
        beforeAll(() => {
            const loggerService = new LoggerService({});
            const gitlabRepository = new GitlabRepository({});
            tagService = new TagService({ gitlabRepository, loggerService, config: {} });
        });
        test("should return true when tag matches regex", () => {
            const result = tagService.isTagsMatchTargetTagRegex([{ name: "1.1.0" }], /^[0-9]+.[0-9]+.[0-9]+$/);
            expect(result).toBeTruthy();
        });
        test("should return false when tag doesn't match regex", () => {
            const result = tagService.isTagsMatchTargetTagRegex([{ name: "1.1.0-1" }], /^[0-9]+.[0-9]+.[0-9]+$/);
            expect(result).toBeFalsy();
        });
        test("should return true when regex is missing", () => {
            const result = tagService.isTagsMatchTargetTagRegex([{ name: "1.1.0-1" }], undefined);
            expect(result).toBeTruthy();
        });
    });
    describe("#isBranchesInTargetBranch", () => {
        let tagService;
        beforeAll(() => {
            const loggerService = new LoggerService({});
            const gitlabRepository = new GitlabRepository({});
            tagService = new TagService({ gitlabRepository, loggerService, config: {} });
        });
        test("should return true when branch contains target branch", () => {
            const result = tagService.isBranchesInTargetBranch([{ name: "master" }], "master");
            expect(result).toBeTruthy();
        });
        test("should return false when branch doesn't contain target branch", () => {
            const result = tagService.isBranchesInTargetBranch([{ name: "develop" }], "master");
            expect(result).toBeFalsy();
        });
        test("should return true when target branch is missing", () => {
            const result = tagService.isBranchesInTargetBranch([{ name: "develop" }], undefined);
            expect(result).toBeTruthy();
        });
    });
});
