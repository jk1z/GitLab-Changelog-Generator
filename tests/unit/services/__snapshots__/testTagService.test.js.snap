// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag service #getLatestAndSecondLatestTag Boom scenario Latest tag is not in the targeting branch should throw err 1`] = `"Latest tag doesn't contain target branch. Target branch develop"`;

exports[`Tag service #getLatestAndSecondLatestTag Boom scenario Latest tag is not matching the regex should throw err 1`] = `"Latest tag doesn't match with the regex. Target tag regex /^[0-9]+.[0-9]+.[0-9]+$/"`;

exports[`Tag service #getLatestAndSecondLatestTag Optional scenario Only one tag should call findBranchRefsByProjectIdAndSha 1`] = `
Array [
  Array [
    "123456",
    "59395c05c18b9c8904853715d4136921de0b48f1",
  ],
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Optional scenario Only one tag should call findRepoByProjectId 1`] = `
Array [
  "123456",
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Optional scenario Only one tag should call findTagsByProjectId 1`] = `
Array [
  "123456",
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Optional scenario Only one tag should log info 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Latest tag is v1.0.0",
    ],
    Array [
      "No more tag is found. Assuming project creation date is the start date",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Tag service #getLatestAndSecondLatestTag Optional scenario Only one tag should return latest and second latest tag 1`] = `
Array [
  Object {
    "commit": Object {
      "author_email": "john@example.com",
      "author_name": "John Smith",
      "authored_date": "2012-05-25T04:42:42Z",
      "committed_date": "2012-05-25T04:42:42Z",
      "committer_email": "jack@example.com",
      "committer_name": "Jack Smith",
      "created_at": "2012-05-25T04:42:42Z",
      "id": "59395c05c18b9c8904853715d4136921de0b48f1",
      "message": "Initial commit",
      "parent_ids": Array [
        "2a4b78934375d7f53875269ffd4f45fd83a84ebd",
      ],
      "short_id": "de0b48f1",
      "title": "Initial commit",
    },
    "message": null,
    "name": "v1.0.0",
    "release": Object {
      "description": "Amazing release. Wow",
      "tag_name": "1.0.0",
    },
    "target": "59395c05c18b9c8904853715d4136921de0b48f1",
  },
  Object {
    "commit": Object {
      "committed_date": "2012-05-26T04:42:42Z",
    },
  },
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Sunny scenario should call findBranchRefsByProjectIdAndSha 1`] = `
Array [
  Array [
    "123456",
    "ec0b4f0b5c90ed0fa911a2972ccc452641b31563",
  ],
  Array [
    "123456",
    "ec0b4f0b5c90ed0fa911a2972ccc452641b31563",
  ],
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Sunny scenario should call findTagsByProjectId 1`] = `
Array [
  "123456",
]
`;

exports[`Tag service #getLatestAndSecondLatestTag Sunny scenario should log info 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Latest tag is v1.1.0",
    ],
    Array [
      "Found the second latest tag on page 1. The second latest tag is v1.0.0",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Tag service #getLatestAndSecondLatestTag Sunny scenario should return latest and second latest tag 1`] = `
Array [
  Object {
    "commit": Object {
      "author_email": "john@example.com",
      "author_name": "John Smith",
      "authored_date": "2012-05-27T04:42:42Z",
      "committed_date": "2012-05-27T04:42:42Z",
      "committer_email": "jack@example.com",
      "committer_name": "Jack Smith",
      "created_at": "2012-05-27T04:42:42Z",
      "id": "ec0b4f0b5c90ed0fa911a2972ccc452641b31563",
      "message": "Initial commit",
      "parent_ids": Array [
        "2a4b78934375d7f53875269ffd4f45fd83a84ebe",
      ],
      "short_id": "41b31563",
      "title": "Initial commit",
    },
    "message": null,
    "name": "v1.1.0",
    "release": Object {
      "description": "Amazing release. Wow",
      "tag_name": "1.1.0",
    },
    "target": "ec0b4f0b5c90ed0fa911a2972ccc452641b31563",
  },
  Object {
    "commit": Object {
      "author_email": "john@example.com",
      "author_name": "John Smith",
      "authored_date": "2012-05-25T04:42:42Z",
      "committed_date": "2012-05-25T04:42:42Z",
      "committer_email": "jack@example.com",
      "committer_name": "Jack Smith",
      "created_at": "2012-05-25T04:42:42Z",
      "id": "59395c05c18b9c8904853715d4136921de0b48f1",
      "message": "Initial commit",
      "parent_ids": Array [
        "2a4b78934375d7f53875269ffd4f45fd83a84ebd",
      ],
      "short_id": "de0b48f1",
      "title": "Initial commit",
    },
    "message": null,
    "name": "v1.0.0",
    "release": Object {
      "description": "Amazing release. Wow",
      "tag_name": "1.0.0",
    },
    "target": "59395c05c18b9c8904853715d4136921de0b48f1",
  },
]
`;
